### DISCUSSION TITLE
Question about using autodifferentiation with gpytorch model sample outputs

### DISCUSSION THREAD

**Discussion by neelshah2397** [cornellius-gp/gpytorch#1542] [Q&A] [answered]
I am trying to minimize a function with respect to x values. To do so, I use the code at the bottom of this post. As I understand it, I should be able to make an initial guess, set the requires_grad flag to true and run the forward pass (scores = alpha(Xsamples, model, robustness)) and then get the gradients with scores.backward() and then update my initial guess accordingly with optimizer.step. 

However, when I try running this I get the following error 'RuntimeError: element 0 of tensors does not require grad and does not have a grad_fn' which I don't understand because I have set my initial guess to requires gradient. I think this is because the sample() function I use to calculate 'scores' applies a torch.no_grad() context manager before computing its output. I tried looking on forums for help but most of the answers were regarding training neural networks so their fixes did not work in this case. Any guidance on this would be greatly appreciated, thank you.

`epoch = 100`
`learning_rate = 0.01`
`N = 1`
`Xsamples = torch.randn(1,2)`
`Xsamples.requires_grad = True`
`optimizer = torch.optim.SGD([Xsamples], lr = learning_rate)`
`for i in range(epoch):`
`   scores = alpha(Xsamples, model, robustness)`
`    scores.backward() #dscore/ dx`
`   optimizer.step()`
`   optimizer.zero_grad()`
`return Xsamples `

**Comment by gpleiss** [comment, marked as answer]
What is `alpha`? Also, you should make `Xsamples` a PyTorch parameter: `Xsamples = torch.nn.Parameter(torch.randn(1, 2))`

**Reply by neelshah2397** [reply]
Hi Geoff,

Thanks for the reply. I will incorporate Xsamples as suggested. Alpha is a function which computes a Monte Carlo estimate which requires sampling from the GP. A problem I found is if we do f = likelihood(model(x)) then f.sample() seems to break the computation graph as the sample() method applies a torch.no_grad() context manager before computing its output.

I avoided this issue by switching it to f.mean but then it seems to just return NaN for the backward step.

**Reply by Balandat** [reply]
you'll want to use `rsample` instead
